yarn-project/p2p/src/mem_pools/tx_pool/eviction/invalid_txs_after_mining_rule.ts

```diff
@@ -0,0 +1,104 @@
+import { createLogger } from '@aztec/foundation/log';
+import type { TxHash } from '@aztec/stdlib/tx';
+
+import type { TxPoolOptions } from '../tx_pool.js';
+import {
+  type EvictionContext,
+  EvictionEvent,
+  type EvictionResult,
+  type EvictionRule,
+  type TxPoolOperations,
+} from './eviction_strategy.js';
+
+/**
+ * Eviction rule that removes invalid transactions after a block is mined.
+ * Only triggers on BLOCK_MINED events.
+ *
+ * Eviction criteria includes:
+ * - Transactions with nullifiers that are already included in the mined block
+ * - Transactions with an expiration timestamp less than or equal to the mined block timestamp
+ */
+export class InvalidTxsAfterMiningRule implements EvictionRule {
+  public readonly name = 'InvalidTxsAfterMining';
+
+  private log = createLogger('p2p:mempool:tx_pool:invalid_txs_after_mining_rule');
+
+  async evict(context: EvictionContext, txPool: TxPoolOperations): Promise<EvictionResult> {
+    if (context.event !== EvictionEvent.BLOCK_MINED) {
+      return {
+        reason: 'block_mined_invalid_txs',
+        success: true,
+        txsEvicted: [],
+      };
+    }
+
+    if (!context.block || !context.newNullifiers) {
+      this.log.warn('Invalid context for block mined eviction', { context });
+      return {
+        reason: 'block_mined_invalid_txs',
+        success: false,
+        txsEvicted: [],
+        error: new Error('Invalid block mined context'),
+      };
+    }
+
+    try {
+      const { timestamp } = context.block.globalVariables;
+
+      const txsToEvict: TxHash[] = [];
+      const pendingTxs = await txPool.getPendingTxInfos();
+      const minedNullifiers = new Set(context.newNullifiers.map(n => n.toString()));
```

Comment:
I was thinking about the performance hit of `Fr.toString`. Maybe we should use the bigint representation instead?

Reply:
Fixed it in I think a better way w/o any conversions (using `sortedArray`)